docs/decisions/0007-handling-learner-initiated-unenrollments.rst

```diff
@@ -0,0 +1,101 @@
+++++++++++++++++++++++++++++++++++++++++++
+
+As it currently stands (May 2021), the enterprise subsidy service is not a part of the Jenkins CI pipeline. The
+accepted implementation of running routine jobs for our services hosted on kubernetes is to utilize argoCD. within
+our internal configurations we can ensure that the job is run on a schedule of our choosing.
+
+Admin initiated unenrollment DAG and introducing a transaction reversal writing task
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+The subsidy service currently utilizes specific event driven architecture to handle admin initiated unenrollments.
+This is done by listening for reversal object creations and then making a call to platform to unenroll all associated
+fulfillment objects. This runs in opposition to the planned management command which will write reversals based on the 
+cancellation of fulfillment objects. In order to mitigate any cyclical behaviors between the two systems, we will need
+to implement two stop-gaps.
+
+- The platform enterprise fulfillment cancellation endpoint needs to verify if the fulfillment object is already
+  revoked and ensure that the unenrolled_at time of the course enrollment object is not bumped if so.
+
+- The subsidy service must check if a recently unenrolled fulfillment object already has an associated reversal object.
+
+How the systems will interact with one another
+++++++++++++++++++++++++++++++++++++++++++++++
+
+Records updated and monitored are limited to the subsidy and edx platform services. Notably, platform course enrollment
+and enterprise fulfillment objects are both housed within edx platform but are split between the enterprise package and
+the LMS. The subsidy service will be responsible for maintaining reversals, as well as their transaction counterparts.
+
+The flow for learner initiated unenrollments will be as follows:
+
+1. Learners unenroll from a course. Notably this has multiple origin sources as it can be done from the LMS or from
+   the enterprise learner/admin dashboards, however the end result is that the course enrollment object on platform is
+   cancelled.
+2. Enterprise enrollment objects will be synchronously updated to reflect the unenrollment of the course enrollment via
+   a Django signal event/handler. 
+3. The subsidy service will routinely hit the enterprise fulfillment recent unenrollment endpoint which will surface
+   enterprise fulfillments tied to recently canceled enterprise enrollments.
+4. The subsidy service will then check for the existence of a reversal object associated with the fulfillment object's
+   transactions. It will then determine if transaction is refundable and if so, write a reversal object. 
+5. After the generation of a transaction reversal, the subsidy service sends a request to platform to unenroll 
+   fulfillment objects which will result in a NOOP as the fulfillments will have already been cancelled.
```

Comment:
praise: thank you for writing this out, makes it much easier to reason about the big picture!